### ISSUE TITLE
Poor performance resulting from audio nodes being created dynamically

### ISSUE THREAD

**Issue by lpil** [christopheralcock/falling3rds#3]
Hey!

@zoeabryant showed me your app. Really cool stuff, love how it sounds.

I noticed that it seems to be using a lot of CPU and struggles with 5 layers playing at once, so I read the source to see how we could make it more performant.

The main problem is that audio nodes are being created dynamically as we iterate and as more layers are added. Creation is these nodes is very expensive, and is to be avoided where possible.

Instead we can create a "initialize" step for our application in which we create all the needed nodes. Our application then just modulates these existing nodes, which is much faster! :)

Cheers,
Louis

**Comment by christopheralcock** [comment]
Hi Louis,

Thanks very much for your time!

My experience of the CPU issues was that on an early version of the code I was using dozens of audiocontexts across multiple browser windows without obvious problem, and it was the adding of a delay effect to the notes that caused the CPU problems, leading me to think that the real CPU issue is that it's changed from chucking a bunch of notes to a synth to effectively manipulating and looping audio. So I built in the limit of 5 layers of sound, which is  too much for some computers definitely, especially after a while.

I'm not sure how to go about implementing your suggestion though - could you clarify exactly what you mean by a node in this context, please?

Cheers,

Chris

**Comment by lpil** [comment]
A node being any of the audio generating or audio transforming objects that you create from the audio context :)

Treat web audio API objects like bits of a synth hardware in a studio or music performance- hook everything up before you start, and then just turn knobs and feed notes into it when you're playing.

**Comment by christopheralcock** [comment]
Hi Louis,
Just saw you close this but I'd like to thank you for your help on this.
I actually solved the issue quite differently to how you suggested, but without your flagging and explanation of the problem I probably wouldn't have got it sorted.
To do what you suggested would have meant restructuring the whole app but instead I set timeouts on the nodes to disconnect, so that when garbage collection comes along it can do its job more efficiently and with no popping.  Not an elegant solution but pretty effective.
Cheers,
Chris
